test/integration/consul-container/test/upgrade/README.md

```diff
@@ -78,34 +117,53 @@ see example [here](./hashicorp/consul/test/integration/consul-container/test/upg
+3. Call the upgrade method and assert the upgrading cluster succeeds.
+We also restart the envoy proxy to make sure the upgraded agent can generate
+the correct envoy configurations.
 
-Then we validate traffic between the static-server and static-client envoy sidecar. After validation, we take a snapshot and upgrade the Consul cluster to the `target-version` and restart the sidecars. Re-validate the client, server and sidecars to ensure the data snapshotted from the pervious version can be accessed in the latest version. 
+```go
+	err = cluster.StandardUpgrade(t, context.Background(), targetVersion)
+	require.NoError(t, err)
+	require.NoError(t, staticServerConnectProxy.Restart())
+	require.NoError(t, staticClientConnectProxy.Restart())
+```
 
+4. Verify the user workload after upgrade, e.g.,
+
+```go
+	libassert.HTTPServiceEchoes(t, "localhost", port, "")
+	libassert.AssertFortioName(t, fmt.Sprintf("http://localhost:%d", appPort), "static-server-2-v2", "")
+```
 
 ### Errors Test Cases
 There are some caveats for special error handling of versions prior to `1.14`.   
 Upgrade tests for features such peering, had API changes that returns an error if attempt to upgrade, and should be accounted for in upgrade tests. If running upgrade tests for any version before `1.14`, the following lines of code needs to be added to skip test or it will not pass.  
 
 ```go
-		fromVersion, err := version.NewVersion(utils.LatestVersion)
-		require.NoError(t, err)
-		if fromVersion.LessThan(utils.Version_1_14) {
-			continue
-		}
+	fromVersion, err := version.NewVersion(utils.LatestVersion)
+	require.NoError(t, err)
+	if fromVersion.LessThan(utils.Version_1_14) {
+		continue
+	}
 ```
 See example [here](https://github.com/hashicorp/consul-enterprise/blob/005a0a92c5f39804cef4ad5c4cd6fd3334b95aa2/test/integration/consul-container/test/upgrade/peering_control_plane_mgw_test.go#L92-L96)
 
-To write tests for bugs found during upgrades, see example on how to add a testcase for those scenarios [here](./hashicorp/consul/test/integration/consul-container/test/upgrade/fullstopupgrade_test.go). 
+To write tests for bugs found during upgrades, see example on how to add a testcase for those scenarios [here](./fullstopupgrade_test.go). 
 
 ## FAQS
 
-**Q.** To troubleshoot, how can I send API request or consul command to the deployed cluster?
+**Q.** Are containers' ports (e.g., consul's 8500, envoy sidecar's admin port
+or local upstream port) are exposed on the docker host? \
+**A.** Yes, they are exposed. However, they are exposed through a [pod container](https://github.com/hashicorp/consul/blob/57e034b74621180861226a01efeb3e9cedc74d3a/test/integration/consul-container/libs/cluster/container.go#L132).
+That is, a consul agent and the envoy proxy containers registered with the agent
+share the [same network](https://github.com/hashicorp/consul/blob/57e034b74621180861226a01efeb3e9cedc74d3a/test/integration/consul-container/libs/cluster/app.go#L23-L30) as the pod container.
```

Comment:
"network" could be confused with "Docker network", which is technically true, but not the point you're trying to make here :)

```suggestion
share the [same Linux network namespace (i.e., they share `localhost`)](https://github.com/hashicorp/consul/blob/57e034b74621180861226a01efeb3e9cedc74d3a/test/integration/consul-container/libs/cluster/app.go#L23-L30) as the pod container.
```